### PR TITLE
feat: create match and add validation

### LINKED ISSUE #9: [ Match Cat ] Create Match
- POST /v1/cat/match
    - Request
    
    ```jsx
    {
    	"matchCatId": "",
    	"userCatId": "",
    	"message": "" // not null, minLength: 5, maxLength: 120
    }
    ```
    
    - Response
        - `200` successfully send match request
        - `404` if neither `matchCatId` / `userCatId` is not found
        - `400` if the cat’s gender is same
        - `400` if neither `matchCatId` / `userCatId` already matched
        - `400` if `matchCatId` / `userCatId` is from the same owner
        - `401` request token is missing or expired

### PR DESCRIPTION
**Pull request by shevaathalla**
Close #9 

### Postman Collection

[Cat Social.postman_collection.json](https://github.com/sedelinggam/cats-social/files/15178655/Cat.Social.postman_collection.json)

#### Note
aku ngubah dikit jd pake environment postman

### Progress
- create match migrations
- create match service n repo
- add validation
image

### CODE REVIEW

File: internal/delivery/http/v1/controller/matches/matches_controller.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+	matchHandler "cats-social/internal/delivery/http/v1/controller/matches/handler"
+	matchesService "cats-social/internal/service/matches"
+	"cats-social/pkg/auth"
+
+	"github.com/gofiber/fiber/v2"
+)
+
+func Init(group fiber.Router, matchSvc matchesService.MatchService) {
+	handler := matchHandler.NewHandler(matchSvc)
+	user := group.Group("/cat/match")
```
Comment (abdmmar) [inline internal/delivery/http/v1/controller/matches/matches_controller.go]:
```suggestion
	match := group.Group("/cat/match")
```

File: internal/delivery/http/v1/request/matches_request.go [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package request
+
+type CreateMatch struct {
+	MatchCatID string `json:"matchCatId" validate:"required"`
+	UserCatID  string `json:"userCatId" validate:"required"`
+	Message    string `json:"message" validate:"required"`
```
Comment (abdmmar) [inline internal/delivery/http/v1/request/matches_request.go]: Ini udah pake `validate` cuman kayaknya belum pake `validator` di `create_match_handler`-nya ya? sama kayaknya bisa tambahin validasi `min` sama `max` char:

```suggestion
	Message    string `json:"message" validate:"required,min=5,max=120"`
```

File: internal/delivery/http/v1/controller/matches/handler/create_match_handler.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+		resp *response.CreateMatch
+		err  error
+	)
+	err = c.BodyParser(&req)
+	if err != nil {
+		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
+			"error": err.Error(),
+		})
+	}
+
```
Comment (abdmmar) [inline internal/delivery/http/v1/controller/matches/handler/create_match_handler.go:30]: Kyknya bisa ditambahin validasi disini:
```suggestion
// Create a new validator instance
	validate := validator.New()

	// Validate the Create Match struct
	err = validate.Struct(req)
	if err != nil {
		// Validation failed, handle the error
		return lumen.FromError(lumen.NewError(lumen.ErrBadRequest, err)).SendResponse(c)
	}
```

**Review by abdmmar — CHANGES_REQUESTED** [review]
Commented, sama boleh tolong attach postman collection-nya di deskripsi? thanks

File: internal/delivery/http/v1/controller/matches/handler/create_match_handler.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+	//Get jwt user ID
+	user := c.Locals("user").(*jwt.Token)
+	claims := user.Claims.(jwt.MapClaims)
+	userID := claims["id"].(string)
+	ctx := c.Context()
+	ctx.SetUserValue("user_id", userID)
+	resp, err = mh.matchService.CreateMatch(ctx, req)
+	if err != nil {
+		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
```
Comment (malikilamalik) [inline internal/delivery/http/v1/controller/matches/handler/create_match_handler.go]: Error handlingnya pak.
```suggestion
		return lumen.FromError(err).SendResponse(c)
```

File: internal/delivery/http/v1/request/matches_request.go (hunk 2 above)
Comment (shevaathalla) [reply internal/delivery/http/v1/request/matches_request.go]: adjusted mas thanks

File: internal/delivery/http/v1/controller/matches/handler/create_match_handler.go (hunk 4 above)
Comment (shevaathalla) [reply internal/delivery/http/v1/controller/matches/handler/create_match_handler.go]: adjusted mas

**Review by abdmmar — APPROVED** [review]

**Review by malikilamalik — APPROVED** [review]

**Comment by abdmmar** [comment]
eh @shevaathalla merge-nya ke `main` branch ya

**Comment by shevaathalla** [comment]
> eh @shevaathalla merge-nya ke `main` branch ya

adjusted

**Review by achmadmuzaki14 — APPROVED** [review]

**Pull request merged**